compiler/rustc_mir_transform/src/simplify_static_switch.rs

```diff
@@ -0,0 +1,165 @@
+        if data.is_cleanup {
+            continue;
+        }
+
+        let predecessors = &predecessors[block];
+        if predecessors.is_empty() {
+            continue;
+        }
+
+        let TerminatorKind::SwitchInt {
+            discr: Operand::Move(switched) | Operand::Copy(switched),
+            ref targets
+        } = data.terminator().kind else {
+            continue
+        };
+
+        let Some(switched) = switched.as_local() else {
+            continue
+        };
+
+        if borrowed_locals.contains(switched) {
+            continue;
+        }
+
+        let Some(discr) = data.statements.iter().rev().find_map(|statement| {
+            match statement.kind {
+                StatementKind::Assign(box (place, Rvalue::Discriminant(discr)))
+                    if place.as_local() == Some(switched)
+                        && let Some(local) = discr.as_local()
+                        && !borrowed_locals.contains(local) =>
+                {
+                    Some(discr)
+                },
+                _ => None,
+            }
+        }) else {
+            continue
+        };
+
+        for &pred in predecessors {
+            let data = &body.basic_blocks[pred];
+            let TerminatorKind::Goto { .. } = data.terminator().kind else {
+                continue
+            };
+
+            for statement in data.statements.iter().rev() {
+                match statement.kind {
+                    StatementKind::Assign(box (
+                        place,
+                        Rvalue::Aggregate(box AggregateKind::Adt(def_id, variant, substs, ..), ..),
+                    )) if place == discr => {
+                        debug!("{pred:?}: {place:?} = {def_id:?}::{variant:?}; goto -> {block:?};");
+                        let discr_ty = tcx.type_of(def_id).subst(tcx, substs);
+                        if let Some(discr) = discr_ty.discriminant_for_variant(tcx, variant) {
+                            let target = targets.target_for_value(discr.val);
+                            new_targets
+                                .entry((place, block))
+                                .or_insert_with(Vec::new)
+                                .push((pred, target));
+                        }
+                    }
+                    _ => {}
+                }
+            }
+        }
+    }
+
+    if new_targets.is_empty() {
+        return false;
+    }
+
+    let basic_blocks = body.basic_blocks.as_mut();
+    for ((place, block), new_targets) in new_targets {
+        let block = &basic_blocks[block];
+        let statements = if block.statements.len() > 1 {
+            block
+                .statements
+                .iter()
+                .filter_map(|statement| match statement.kind {
+                    StatementKind::Assign(box (.., Rvalue::Discriminant(discr)))
+                        if place == discr =>
+                    {
+                        None
```

Comment:
Why remove this statement? The LHS (`switched`) may still be used, even if we remove the `SwitchInt`.

Reply:
now all of the statements are cloned instead of ignoring this statement.